src/sortedarray.c

```diff
@@ -0,0 +1,273 @@
+                                  unsigned int right)
+{
+	unsigned int index = right;
+
+	while (left < right) {
+		index = (left + right) / 2;
+
+		int order = sortedarray->cmp_func(data, 
+		                                  sortedarray->data[index]);
+		if (order <= 0) {
+			left = index + 1;
+		}
+		else {
+			right = index;
+		}
+	}
+
+	return index;
+}
+
+SortedArray *sortedarray_new(unsigned int length,
+                             SortedArrayEqualFunc equ_func,
+                             SortedArrayCompareFunc cmp_func)
+{
+	/* check input requirements */
+	if (equ_func == NULL || cmp_func == NULL) {
+		return NULL;
+	}
+
+	/* If length is 0, set it to a default. */
+	if (length == 0) {
+		length = 16;
+	}
+
+	SortedArrayValue *array = malloc(sizeof(SortedArrayValue) * length);
+
+	/* on failure, return null */
+	if (array == NULL) {
+		return NULL;
+	}
+
+	SortedArray *sortedarray = malloc(sizeof(SortedArray));    
+
+	/* check for failure */
+	if (sortedarray == NULL) {
+		free(array);
+		return NULL;
+	}
+    
+	/* init */
+	sortedarray->data = array;
+	sortedarray->length = 0;
+	sortedarray->_alloced = length;
+	sortedarray->equ_func = equ_func;
+	sortedarray->cmp_func = cmp_func;
+	return sortedarray;
+}
+
+void sortedarray_free(SortedArray *sortedarray)
+{
+	if (sortedarray != NULL) {
+		free(sortedarray->data);
+		free(sortedarray);
+	}
+}
+
+void sortedarray_remove(SortedArray *sortedarray, unsigned int index)
+{
+	/* same as remove range of length 1 */
+	sortedarray_remove_range(sortedarray, index, 1);
+}
+
+void sortedarray_remove_range(SortedArray *sortedarray, unsigned int index,
+                              unsigned int length)
+{
+	/* removal does not violate sorted property */
+
+	/* check if valid range */
+	if (index > sortedarray->length || index + length > sortedarray->length) {
+		return;
+	}
+
+	/* move entries back */
+	memmove(&sortedarray->data[index],
+	        &sortedarray->data[index + length],
+	        (sortedarray->length - (index + length)) 
+	              * sizeof(SortedArrayValue));
+
+	sortedarray->length -= length;
+}
+
+int sortedarray_insert(SortedArray *sortedarray, SortedArrayValue data)
+{
+	/* do a binary search like loop to find right position */
+	unsigned int left  = 0;
+	unsigned int right = sortedarray->length;
+	unsigned int index = 0;
+
+	right = (right > 1) ? right : 0;
+
+	while (left != right) {
+		index = (left + right) / 2;
+
+		int order = sortedarray->cmp_func(data, 
+		                                  sortedarray->data[index]);
+		if (order < 0) {
+			/* value should be left of index */
+			right = index;
+		}
+		else if (order > 0) {
+			/* value should be right of index */
+			left = index + 1;
+		}
+		else {
+			/* value should be at index */
+			left = right = index;
+			break;
+		}
+	}
+
+	if (sortedarray->length > 0 && sortedarray->cmp_func(data, sortedarray->data[index]) > 0) {
+		index++;
+	}
+
+	/* insert element at r */
```

Comment:
Comment ("at r") is wrong. Can you factor out this block into a separate function, eg.  "sortedarray_grow_to_minimum_size"?